### ISSUE TITLE
[Bug] MacOs overlay/overlay2 sendfile syscall in Alpine Linux

### ISSUE THREAD

**Issue by dpchamps** [moby/moby#42998]
**Description**

Under certain circumstances MacOS `overlay` and `overlay2` storage drivers cause the syscalls `sendfile` and `copy_file_range` to behave unexpectedly in Alpine Linux images (and sometimes other distros).

This has at least propagated into one known bug outside of docker land: [see nodejs copy_file* infinite loops](https://github.com/nodejs/node/issues/40200).

A file apparently gets into a bad state after a `sendfile` syscall performs an operation on it, which can be described as:

* a file (`file-copy`) has data copied to it from another file using [sendfile](https://man7.org/linux/man-pages/man2/sendfile.2.html)
* a new file get data copied to it from `file-copy`  via [copy_file_range](https://man7.org/linux/man-pages/man2/copy_file_range.2.html), or a subsequent `sendfile`

Debugging the overlay storage drivers is out of my skillset. Copied files appear to be identical, nothing looks off with newly created files (stat, ls -ali, diff, hexdump all give results that one would expect to see).

**Steps to reproduce the issue:**
See: https://github.com/dpchamps/sendfile-overlay-bug-repro

### Less precise way
(captured here: https://github.com/dpchamps/sendfile-overlay-bug-repro/blob/main/do-test.sh)
1. pull from any `alpine`
1. in the image, create a file, `cp` that file to a new file
1. execute any command that uses `copy_file_range` and watch it fail

### More precise way
(captured here: https://github.com/dpchamps/sendfile-overlay-bug-repro/blob/main/sendfile-repro.c)
1. Create a file, copy data to it via sendfile
2. Create another file, attempt to copy data from the file created in step 1 with `sendfile` or `copy_file_range`, see zero bytes. 

I've published these to the public registry for ease of reproducing:

* `dchampz/copy_file_range-test:alpine-312` [Dockerfile](https://github.com/dpchamps/sendfile-overlay-bug-repro/blob/main/dockerfiles/Dockerfile-Alpine312)
* `dchampz/copy_file_range-test:alpine-312-sendfileRepro` [Dockerfile](https://github.com/dpchamps/sendfile-overlay-bug-repro/blob/main/dockerfiles/Dockerfile-Alpine312-sendfileRepro)

**Describe the results you received:**

`copy_file_range` and `sendfile` do not behave as expected

**Describe the results you expected:**

For those syscalls to behave as expected.

**Additional information you deem important (e.g. issue happens only occasionally):**

I managed to isolate this problem to exactly MacOS, for the `overlay` and `overlay2` drivers. All of these examples work as expected outside of MacOs (as far as I know), or with docker engine configured to use `vfs` storage driver.

Interestingly, `dchampz/copy_file_range-test:debian-11` works as expected. `dchampz/copy_file_range-test:debian-11-sendfileRepro` exhibits this same bug-like behavior on MacOs Docker, but not Linux Debian Docker.

**Output of `docker version`:**

```
Docker version 20.10.8, build 3967b7d
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.6.3)
  compose: Docker Compose (Docker Inc., v2.0.0)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 33
  Running: 0
  Paused: 0
  Stopped: 33
 Images: 30
 Server Version: 20.10.8
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b63
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.10.47-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 1.938GiB
 Name: docker-desktop
 ID: 7HJY:R2IX:PVSX:6XBA:SV7L:MAUC:BB3U:FQUR:4444:33U4:VKXH:WOOJ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
N/A

**Comment by thaJeztah** [comment]
Wondering if this is related to https://github.com/moby/moby/pull/42681 / https://github.com/moby/moby/pull/42836, which is included in docker engine 20.10.10 (which is now available as part of the latest Docker Desktop 4.2).


I'm not sure _exactly_ what I need to be looking for in the output of your reproduction images, so perhaps you could give it a try after updating to the latest Docker Desktop?

This is the output I got from running the images mentioned in your repository;


```bash
docker run --rm dchampz/copy_file_range-test:alpine-312

Create a new file, copy it (with cp) and then attempt to make copies with copy_file_range.
  * A successful run results in four files, all with the same size.
  * A repro produces four files, where testb-copy has zero bytes written to it.

total 8
 141901 -rw-r--r--    1 root     root             6 Nov 10 12:31 testa
 141902 -rw-r--r--    1 root     root             6 Nov 10 12:31 testb

Bytes written 6
Bytes written 0

total 12
 141901 -rw-r--r--    1 root     root             6 Nov 10 12:31 testa
 141903 -rw-r--r--    1 root     root             6 Nov 10 12:31 testa-copy
 141902 -rw-r--r--    1 root     root             6 Nov 10 12:31 testb
 141904 -rw-r--r--    1 root     root             0 Nov 10 12:31 testb-copy
```

```bash
docker run --rm dchampz/copy_file_range-test:alpine-312-sendfileRepro

	sendfile bytes written: 16
	sendfile -> copy_file_range bytes written: 0
	sendfile -> sendfile bytes written: 0
	copy_file_range bytes written: 16
	cfr->cfr bytes written: 16
```

```bash
docker run --rm dchampz/copy_file_range-test:debian-11-sendFile

	sendfile bytes written: 16
	sendfile -> copy_file_range bytes written: 0
	sendfile -> sendfile bytes written: 0
	copy_file_range bytes written: 16
	cfr->cfr bytes written: 16
```

```bash
docker run --rm dchampz/copy_file_range-test:debian-11


Create a new file, copy it (with cp) and then attempt to make copies with copy_file_range.
  * A successful run results in four files, all with the same size.
  * A repro produces four files, where testb-copy has zero bytes written to it.

total 8
142027 -rw-r--r-- 1 root root 6 Nov 10 12:35 testa
142028 -rw-r--r-- 1 root root 6 Nov 10 12:35 testb

Bytes written 6
Bytes written 6

total 16
142027 -rw-r--r-- 1 root root 6 Nov 10 12:35 testa
142029 -rw-r--r-- 1 root root 6 Nov 10 12:35 testa-copy
142028 -rw-r--r-- 1 root root 6 Nov 10 12:35 testb
142030 -rw-r--r-- 1 root root 6 Nov 10 12:35 testb-copy
```

**Comment by dpchamps** [comment]
Hey @thaJeztah, thanks for responding. Sorry, the output is dense. This is just such a bizarre bug, by the time I'd gotten to the repros I was already far in the weeds.

Looking at the output of the programs you pasted, you've triggered the failure case in the first three: zero bytes written for a copy of a copy. So you are experiencing the bug. All examples should look like the last program you posted: nothing should have zero bytes sent.

Happy to rerun under latest, but _im guessing_ this is still an issue looking at the output.

This is kind of a slippery problem to state very concisely-- lmk if theres anythi ng I can update to make it more clear.

Essentially, something fails when a file is copied. 
But only when the the syscall `sendfile` was used -- which is what `cp` uses at some point under the hood.

**Comment by thaJeztah** [comment]
Thanks for the extra information, so the faulty cases are the ones with `Bytes written 0`, clear!

Reason I tried on the latest Desktop was that I was wondering (as mentioned) if there was a relation to https://github.com/moby/moby/pull/42681. While that syscall _seems_ unrelated, I've seen reports of things failing for obscure reasons (but they turned out to be related to that syscall). 

In doing so, I completely went past your comment about it only being reproducible on overlay (facepalm :

> I managed to isolate this problem to exactly MacOS, for the `overlay` and `overlay2` drivers. All of these examples work as expected outside of MacOs (as far as I know), or with docker engine configured to use `vfs` storage driver.

OverlayFS being involved here could definitely be an issue yes; there have been "odd" things in the past, and some tools not working well with overlay, but most of those have been resolved over time. Perhaps this is one remaining (I'd have to do a bit of Google searching to see if there's other issues mentioning this). If it's overlay, then kernel version / variant can make a difference.

If you have the chance to do so, it may be worth trying if the issue is reproducible directly on the host, by creating an overlay mount (lowerdir, upperdir, workdir), and running the reproduction steps on files on that mount. I don't think docker does a lot of "special" things for the overlay mounts, but it does set some options (main part of the code is around this area: https://github.com/moby/moby/blob/master/daemon/graphdriver/overlay2/overlay.go#L523-L524)

> But only when the the syscall `sendfile` was used -- which is what `cp` uses at some point under the hood.

Just to be sure, I checked if any of the `sendfile ` syscalls could be blocked by the seccomp profile, but that doesn't appear to be the case; https://github.com/moby/moby/blob/2480bebf59c25991ef88ba5229c7a2e65237510c/profiles/seccomp/default.json#L303-L304
